src/rpc/handlers/AMMInfo.h

```diff
@@ -0,0 +1,125 @@
+#include "rpc/BookChangesHelper.h"
+#include "rpc/RPCHelpers.h"
+#include "rpc/common/MetaProcessors.h"
+#include "rpc/common/Types.h"
+#include "rpc/common/Validators.h"
+
+namespace rpc {
+
+/**
+ * @brief AMMInfoHandler returns information about AMM pools.
+ *
+ * For more info see: https://xrpl.org/amm_info.html
+ */
+class AMMInfoHandler {
+    std::shared_ptr<BackendInterface> sharedPtrBackend_;
+
+public:
+    struct Output {
+        // todo: use better type than json types
+        boost::json::value amount1;
+        boost::json::value amount2;
+        boost::json::value lpToken;
+        boost::json::array voteSlots;
+        boost::json::value auctionSlot;
+        std::uint16_t tradingFee = 0;
+        std::string ammAccount;
+        std::optional<bool> asset1Frozen;
+        std::optional<bool> asset2Frozen;
+
+        uint32_t ledgerIndex = 0;
+        bool validated = true;
+    };
+
+    struct Input {
+        std::optional<ripple::AccountID> accountID;
+        std::optional<ripple::AccountID> ammAccount;
+        ripple::Issue issue1 = ripple::noIssue();
+        ripple::Issue issue2 = ripple::noIssue();
+        std::optional<std::string> ledgerHash;
+        std::optional<uint32_t> ledgerIndex;
+    };
+
+    using Result = HandlerReturnType<Output>;
+
+    AMMInfoHandler(std::shared_ptr<BackendInterface> const& sharedPtrBackend) : sharedPtrBackend_(sharedPtrBackend)
+    {
+    }
+
+    static RpcSpecConstRef
+    spec([[maybe_unused]] uint32_t apiVersion)
+    {
+        static auto const stringIssueValidator =
+            validation::CustomValidator{[](boost::json::value const& value, std::string_view /* key */) -> MaybeError {
+                ASSERT(value.is_string(), "Issue expected to be a string");
+
+                try {
+                    ripple::issueFromJson(value.as_string().c_str());
+                } catch (std::runtime_error const&) {
+                    return Error{Status{RippledError::rpcISSUE_MALFORMED}};
+                }
+
+                return MaybeError{};
+            }};
+
+        static auto const rpcSpec = RpcSpec{
+            {JS(ledger_hash), validation::Uint256HexStringValidator},
+            {JS(ledger_index), validation::LedgerIndexValidator},
+            {JS(asset),
+             meta::WithCustomError{
+                 validation::Type<std::string, boost::json::object>{}, Status(RippledError::rpcISSUE_MALFORMED)
+             },
+             meta::IfType<std::string>{stringIssueValidator},
+             meta::IfType<boost::json::object>{
+                 meta::WithCustomError{validation::AMMAssetValidator, Status(RippledError::rpcISSUE_MALFORMED)},
+             }},
+            {JS(asset2),
+             meta::WithCustomError{
+                 validation::Type<std::string, boost::json::object>{}, Status(RippledError::rpcISSUE_MALFORMED)
+             },
+             meta::IfType<std::string>{stringIssueValidator},
+             meta::IfType<boost::json::object>{
```

Comment:
If we use meta::IfType<boost::json::value>, maybe we don't need the std::string validator?